src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,147 @@
+ *
+ *   final_speed - The final speed at the end of the trajectory [m/s]
+ */
+typedef struct FirmwareRobotPathParameters
+{
+    Polynomial2dOrder3_t path;
+    float t_start;
+    float t_end;
+    unsigned int num_segments;
+    float max_allowable_acceleration;
+    float max_allowable_speed;
+    float initial_speed;
+    float final_speed;
+} FirmwareRobotPathParameters_t;
+
+/**
+ * Returns a planned trajectory as a list of TrajectoryElement's.
+ *
+ *  Key assumptions & guarantees of this planner are:
+ *  - Trajectories are time-optimal assuming INFINITE JERK capability of the robot
+ *  - No speed on the trajectory is larger than the 'max_allowable_speed' parameter.
+ *  - No acceleration value between points on the trajectory can be larger than the
+ * 'max_allowable_acceleration' input parameter
+ *  - Assuming the grip-limit of the robot IS THE SAME AS THE MAX ACCELERATION then at no
+ * point on the path can the sum of centripetal and acceleration force be greater than
+ * 'max_allowable_acceleration'
+ *
+ *  Trajectory generation is done by assuming constant acceleration capability.
+ *      - The generator will assume maximum acceleration for the robot between each
+ * segment on the path. If using max acceleration breaches either the speed limit of the
+ * robot, or the centripetal acceleration limit defined by the curvature and robot speed -
+ * the planner will assume the max acceleration that will remain bellow this limit.
+ *
+ *     - To ensure the robot is capable of deccelerating to reach ant sudden changes in
+ * the path, the trajectory is checked for backwards continuity
+ *
+ * @pre num_segments > 2
+ * @pre max_allowable_acceleration > 0
+ * @pre max_allowable_speed > 0
+ * @pre init_speed >= 0
+ * @pre final_speed >= 0
+ *
+ *  @param path_parameters [in] The data structure including important path parameters
+ * such as: *path - The 3D polynomial representing the geometric path
+ */
+void app_trajectory_planner_generate_constant_arc_length_segmentation(
+    FirmwareRobotPathParameters_t path_parameters, Trajectory_t* trajectory);
+
+/**
+ * Returns a constant interpolation period (time) trajectory based on an input trajectory
+ *
+ * This function is intended to take a variable time trajectory (which is easier to
+ * calculate) then interpolate a constant period trajectory. This is valuable for discrete
+ * time motion controllers that operate only at constant delta-time intervals
+ *
+ * @pre variable_time_trajectory is a valid (obeys the physical limitation of the robot)
+ * trajectory
+ *
+ * @param constant_period_trajectory [in] The array to be filled with the constant time
+ * interpolation period equivalent of the input trajectory
+ * @param variable_time_trajectory [in] The valid trajectory used as
+ * reference for a constant interpolation period trajectory
+ * @param interpolation_period [in] The constant change in time that corresponds to each
+ * trajectory segment
```

Comment:
`constant_period_trajectory` should be an `[out]`  and should be at the end of the list of arguments I think?

Also we should have a `@pre` that dictates that `constant_period_trajectory` is pre-allocated to the maximum trajectory length.

As a thought, might be useful to make macros for trajectory creation and destruction to handle the creation of a `static` array, etc.

Reply:
Made the changes to [out] and the array pre-allocation @pre